execution_workflows/DaPars2/Dapars2_snakemake/workflow/Snakefile

```diff
@@ -0,0 +1,287 @@
+            pd.Series(samples.loc[wildcards.sample, "bam"]).values
+    output:
+        out=os.path.join(config["out_dir"], "{sample}_Chr_Body.bam")
+    log:
+        os.path.join(config["local_log"], "addChrToBam_2")
+    shell:
+        "samtools view {input.bam} | awk 'BEGIN {OFS="\t"} { print $1,$2,"chr"$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15 }' >>{output.out}"
+
+rule addChrToBam_3:
+    """This is the final step of adding 'chr' prefix to bam file. This process concatenate the header and the body of the previous bam files.
+    """
+    input:
+        header=os.path.join(config["out_dir"], "{sample}_Chr_Header.bam"),
+        body=os.path.join(config["out_dir"], "{sample}_Chr_Body.bam")
+    output:
+        out=os.path.join(config["out_dir"], "{sample}_Chr.bam")
+    log:
+        os.path.join(config["local_log"], "addChrToBam_3")
+    shell:
+        # /bin/cat: Argument list too long ----need to fix
+        "cat $(samtools view -H {input.header}) $(samtools view {input.body}) >{output.out}"
+
+rule sortBam:
+    """This step sorts bam files with 'chr'.
+    """
+    input:
+        bam=os.path.join(config["out_dir"], "{sample}_Chr.bam")
+    output:
+        out=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam")
+    log:
+        os.path.join(config["local_log"], "sortBam")
+    threads:
+        4
+    shell:
+        "samtools sort -@ {threads} -o {output.out} {input.bam}"
+
+rule indexBam:
+    """This step indexes sorted bam files.
+    """
+    input:
+        bam=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam")
+    output:
+        out=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam.bai")
+    log:
+        os.path.join(config["local_log"], "indexBam")
+    threads:
+        4
+    shell:
+        "samtools index -b -@ {threads} {input.bam}"
+
+rule generateBedgraph:
+    """This step generates bedgraph files from sample bams.
+    """
+    input:
+        bam=os.path.join(config["out_dir"], "{smaple}_Chr_sorted.bam")
+    output:
+        out=os.path.join(config["out_dir"], "{sample}_Chr.bedgraph")
+    log:
+        os.path.join(config["local_log"], "generateBedgraph")
+    shell:
+        "genomeCoverageBed -bg -ibam {input.bam} >{output.out}"
+
+rule generatePathReadcounts:
+    """This step counts the aligned read of each sample and write  the abspaths and the numbers to a txt.
+    """
+    input:
+        bam=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam")
+    output:
+        out=os.path.join(config["out_dir"], "pathReadcounts.tsv")
+    params:
+        filename=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam")
+    log:
+        os.path.join(config["local_log"],"generatePathReadcounts")
+    shell:
+        # Output: sample path and its sequencing depth with tab separated ----need to fix
+        "echo "$(cd "$(dirname {params.filename})" && pwd -P)/$(basename {params.filename})"$'\t'$(samtools view -c -F 260 {input.bam}) >>{output.out}"
```

Comment:
This is some next level stuff :)) All the double quotes alone make my head spin.

Two possible solutions to make this at least a little easier to debug:
1. Write a small script and execute that instead. Then there's no need to use a one-liner, and you can easily test the script itself, outside of the context of the workflow.
2. Use the `run` directive instead of the `shell` directive and use Python code.

I'm not entirely sure about the implications though, it might be that this will make it harder to dockerize the workflow later on, I don't have much experience with Docker. In that respect, I'm also a little concerned about absolute paths here: this may also be a problem to run this in the cloud. What do you need this rule for? I mean, I know what the rule itself does, but in what context is it used?

Reply:
Thanks for the suggestion:) sorry for the flying double quotes
About the two solutions:
1.  I had tested it in a simple shell script and it worked. When I tested it using snakemake, it failed (EOL while scanning string literal). 
I found that `dirname` and `basename` take a string as input, but in snakemake rule, the {input} is not passed as a string. So I tried to set params to pass a sting to them. But it is still not working. Maybe all the double quotes keep driving snakemake crazy. I also tried to split the code into two rules. Try to get all the paths and reads separately and then concatenate them column-wise. But I afraid that when snakemake runs the rules, the order of the sample and the read counts may mix up. 
2.  I think this method may work. I'll try pysam to see if things can be easier.

This output will be used for sequencing depth normalization. In the configure file of DaPars2, this file will be assigned like this: `sequencing_depth_file=mapping_bam_location_with_depth.txt`. I am not familiar with Docker, either. I am not sure if changing it to a relative path will work in the cloud.

Reply:
Hey just butting myself into conversation with a few suggestions that may help you. Happy to assist directly with code if you'd like, trying to see where I can join in :)

- If you're using multiline/ a long one-liner in a `shell` invocation, I like to wrap in triple quotes to make it a little easier to read. e.g.

```
shell:
    """
    command a {input} | \
    command b | \
    command c > {output}
    """
```

For my next suggestion, my TL; DR is:
- **I think counting mapped reads for each sample and generating table for all samples should be ideally separated, or the rule structured so that it 'aggregates'/runs all samples at once.**


I'm a little wary about how this rule is structured and you may run into problems if the workflow is interrupted. (If my understanding of snakemake is right here) Snakemake only tracks the generation of output files to see if a rule has been completed. This file will be created once a single sample has been passed to the rule (and then updated for each successive sample). If the workflow was interrupted at this point but one sample has passed successfully through this rule, then Snakemake will see `os.path.join(config["out_dir"], "pathReadcounts.tsv")` has been generated and may consider it complete, without all samples being added to the TSV. Should test this with some dry runs though...


I do think this would be more safely structured using `expand` and some Python code or shell/Python script as suggested by @uniqueg though.  `expand` will return a list of all paths to sample names, you can iterate over this in `run` or shell with calls to samtools/pysam to get the mapped read counts. Pseudo-coding this out it could be something like:

```
rule generatePathReadcounts:
    """This step counts the aligned read of each sample and write  the abspaths and the numbers to a txt.
    """
    input:
        bams = expand(os.path.join(config["out_dir"],"") + "{sample}_Chr_sorted.bam", sample = samples.index.tolist())
    output:
        out=os.path.join(config["out_dir"], "pathReadcounts.tsv")
    params:
        filename=os.path.join(config["out_dir"], "{sample}_Chr_sorted.bam")
    log:
        os.path.join(config["local_log"],"generatePathReadcounts")
    run:
        import pysam
        out_list = []
        for bam_path in input.bams:
            bam = pysam.AlignmentFile(bam_path, "rb")
            ## may be a better may to ensure count mapped reads...
            mapped_reads = bam.count(until_eof = True, read_callback = "all")
            bam.close()
            outlist.append("\t".join([bam_path, str(mapped_reads)])
            
        with open(output.out, "w") as outfile:
              outfile.write("\n".join(out_list))
```
Downside is you're running multiple samtools/pysam calls within the same rule instead of them being parallelised by sample across different jobs. If `samtools view -c` is  quick pretty quick then not so much of an issue. The alternative is to have your samtools command as a separate rule, then some quick Python code here to check there are counts for each sample and make the TSV.

More than happy to help with writing this out if you're willing @pinjouwu9325, may need some coordination to prevent duplication though!

Reply:
I  want to say that I also currently work on refactoring some parts to help with the execution, at least locally. I still need some time but if I manage I will push the progress.

Reply:
Thank you guys for the support:) very appreciated

I just update this part which I re-wrote it using `pysam` and almost the same idea as yours @SamBryce-Smith 
And you are right. Since `samtools view -c` will read the whole Sam/Bam file for counting, it may take a while to process.
I was struggling with the {input} data structure because the first for loop did not load them one by one. It printed the whole list of samples (not sure it is a list or not) in one loop. So I tried to test it with the second for loop and it worked. Hope this will work fine in the workflow:) and pysam make things much easier

I am going to work on the output of the mainDapars2. But I found there are still bugs to fix. If you are free now and would like to help, would you help me have a look at `ln101` https://github.com/iRNA-COSI/APAeval/blob/7e51ccf6d3a68c956bc9dc354bc6c4144c04d0bc/execution_workflows/DaPars2/Dapars2_snakemake/workflow/Snakefile#L101   
I found that `cat` won't take too many arguments. So I am thinking of using pysam again to add 'chr' to chromosome name, and combining rules: `addChrToBam_1`, `addChrToBam_2` and `addChrToBam_3`

Thanks again to all of your help:)